Platform/OpenNtfsDxe/Info.c

```diff
@@ -0,0 +1,238 @@
+/** @file
+  Copyright (c) 2022, Mikhail Krichanov. All rights reserved.
+  SPDX-License-Identifier: BSD-3-Clause
+
+  Functional and structural descriptions follow NTFS Documentation
+  by Richard Russon and Yuval Fledel
+**/
+
+#include "NTFS.h"
+#include "Helper.h"
+
+extern UINTN mFileRecordSize;
+
+STATIC
+EFI_STATUS
+GetLabel (
+  IN     EFI_FS   *FileSystem,
+  IN OUT CHAR16   **Label
+  )
+{
+  EFI_STATUS       Status;
+  NTFS_FILE        *BaseMftRecord  = NULL;
+  ATTR_HEADER_RES  *Res;
+
+  ASSERT (FileSystem != NULL);
+  ASSERT (Label != NULL);
+
+  *Label = NULL;
+
+  Status = FsHelpFindFile (L"/$Volume", FileSystem->RootIndex, &BaseMftRecord, FSHELP_REG);
+  if (EFI_ERROR (Status)) {
+    DEBUG ((DEBUG_INFO, "NTFS: Could not find $Volume file.\n"));
+    return Status;
+  }
+
+  if (!BaseMftRecord->InodeRead) {
+    BaseMftRecord->FileRecord = AllocateZeroPool (mFileRecordSize);
+    if (BaseMftRecord->FileRecord == NULL) {
+      DEBUG ((DEBUG_INFO, "Failed to allocate buffer for FileRecord.\n"));
+      FreeFile (BaseMftRecord);
+      FreePool (BaseMftRecord);
+      return EFI_OUT_OF_RESOURCES;
+    }
+
+    Status = ReadMftRecord (BaseMftRecord->File, BaseMftRecord->FileRecord, BaseMftRecord->Inode);
+    if (EFI_ERROR (Status)) {
+      DEBUG ((DEBUG_INFO, "NTFS: Could not read Mft Record #%d.\n", BaseMftRecord->Inode));
+      FreeFile (BaseMftRecord);
+      FreePool (BaseMftRecord);
+      return Status;
+    }
+  }
+
+  Status = InitAttr (&BaseMftRecord->Attr, BaseMftRecord);
+  if (EFI_ERROR (Status)) {
+    DEBUG ((DEBUG_INFO, "NTFS: Could not initialize attribute.\n"));
+    FreeFile (BaseMftRecord);
+    FreePool (BaseMftRecord);
+    return Status;
+  }
+
+  Res = (ATTR_HEADER_RES *) FindAttr (&BaseMftRecord->Attr, AT_VOLUME_NAME);
+  if ((Res != NULL) && (Res->NonResFlag == 0) && (Res->InfoLength != 0)) {
+    //
+    // The Volume Name is not terminated with a Unicode NULL.
+    // Its name's length is the size of the attribute as stored in the header.
+    //
+    *Label = AllocateZeroPool (Res->InfoLength + sizeof (CHAR16));
```

Comment:
We are checking ``Res->InfoLength`` for the first time here? Why can't the addition overflow?